tests/mock_crust/tunnel.rs

```diff
@@ -0,0 +1,148 @@
+// bound by the terms of the MaidSafe Contributor Agreement, version 1.1.  This, along with the
+// Licenses can be found in the root directory of this project at LICENSE, COPYING and CONTRIBUTOR.
+//
+// Unless required by applicable law or agreed to in writing, the SAFE Network Software distributed
+// under the GPL Licence is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.
+//
+// Please review the Licences for the specific language governing permissions and limitations
+// relating to use of the SAFE Network Software.
+
+use itertools::Itertools;
+use routing::{XOR_NAME_LEN, XorName};
+use routing::mock_crust::{Config, Endpoint, Network};
+use super::{TestNode, create_connected_nodes, create_connected_nodes_until_split_with_nodes,
+            poll_all, poll_and_resend, verify_invariant_for_all_nodes};
+
+#[test]
+fn failing_connections_ring() {
+    let min_section_size = 8;
+    let network = Network::new(min_section_size, None);
+    let len = min_section_size * 2;
+    for i in 0..(len - 1) {
+        let ep0 = Endpoint(1 + i);
+        let ep1 = Endpoint(1 + (i % len));
+
+        network.block_connection(ep0, ep1);
+        network.block_connection(ep1, ep0);
+    }
+    let nodes = create_connected_nodes(&network, len);
+    verify_invariant_for_all_nodes(&nodes);
+}
+
+#[test]
+fn failing_connections_bidirectional() {
+    let min_section_size = 4;
+    let network = Network::new(min_section_size, None);
+    network.block_connection(Endpoint(2), Endpoint(3));
+    network.block_connection(Endpoint(3), Endpoint(2));
+
+    let nodes = create_connected_nodes(&network, min_section_size);
+    verify_invariant_for_all_nodes(&nodes);
+}
+
+#[test]
+fn failing_connections_unidirectional() {
+    let min_section_size = 8;
+    let network = Network::new(min_section_size, None);
+    network.block_connection(Endpoint(1), Endpoint(6));
+    network.block_connection(Endpoint(1), Endpoint(7));
+    network.block_connection(Endpoint(6), Endpoint(7));
+
+    let nodes = create_connected_nodes(&network, min_section_size);
+    verify_invariant_for_all_nodes(&nodes);
+}
+
+// Removes nodes from the specified section so that this section will merge with another section.
+fn remove_nodes_from_section_till_merge(prefix_name: &XorName,
+                                        nodes: &mut Vec<TestNode>,
+                                        min_section_size: usize) {
+    let section_indexes: Vec<usize> = nodes.iter()
+        .enumerate()
+        .rev()
+        .filter_map(|(index, node)| if node.routing_table().our_prefix().matches(prefix_name) {
+            Some(index)
+        } else {
+            None
+        })
+        .collect();
+    section_indexes.iter()
+        .take(section_indexes.len() - min_section_size + 1)
+        .foreach(|index| { let _ = nodes.remove(*index); });
```

Comment:
I guess so. I don't much like this style of coding because it's easy to accidentally break it, but I guess it'll do.